### PR TITLE
Update tflops calc for qwen3 next & GatedDeltaNet

### PR DESCRIPTION
**Pull request by Rohan-Bierneni**
# Description

### Description
This PR implements the TFLOPs calculation logic for the Qwen3-Next architecture. Qwen3-Next utilizes a hybrid design containing both standard full attention layers and linear attention (Gated Delta Net) layers, alongside a Mixture-of-Experts (MoE) structure. This update ensures training TFLOPs are accurately reported for this model family.

### Key Changes

**1. TFLOPs Calculation Logic (`src/MaxText/maxtext_utils.py`)**
* **Added `calculate_gated_delta_net_flops_per_device`**: Implemented logic to calculate FLOPs for the Linear Attention layers, breaking down operations into:
    * **Projections**: QKVZ, BA, and Output projections.
    * **Convolution**: Depthwise convolutions on the key/value/gate states.
    * **Core Attention**: Intra-chunk and inter-chunk recurrent state operations.
* **Updated `calculate_tflops_training_per_device`**:
    * Added a specific branch for `DecoderBlockType.QWEN3_NEXT`.
    * Logic now calculates the number of "full attention" layers vs. "linear attention" layers based on the `inhomogeneous_layer_cycle_interval`.
    * Combines FLOPs from embeddings, MoE FFNs (routed + shared), full causal attention, and linear attention.
* **Updated FFN Helpers**: Verified `calculate_routed_and_shared_ffn_tflops_per_device` and `get_dense_moe_layers` explicitly support `QWEN3_NEXT`.

**2. Unit Tests (`tests/unit/flop_calculation_test.py`)**
* **Added `test_qwen3_next_flops`**: A new unit test that verifies the implementation against a "golden" manual calculation for the 80B model configuration.
* **Added `compute_qwen3_next_attention_flops_per_device`**: A helper function to compute the expected attention-specific FLOPs for the test assertion.

**3. Config Updates (`src/MaxText/configs/models/qwen3-next-80b-a3b.yml`)**
* Added `shared_experts: 1` to the model configuration to ensure correct parameter counting in the FFN FLOPs calculation.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/477291633

# Tests

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/maxtext_utils.py [hunk 1]
```diff
@@ -389,6 +389,77 @@ def get_dense_moe_layers(config):
   return num_dense_layers, num_moe_layers
 
 
+def calculate_qwen3_next_ffn_tflops_per_device(config):
```
Comment (RissyRan) [inline src/MaxText/maxtext_utils.py]: Do you think we could directly use calculate_routed_and_shared_ffn_tflops_per_device()? the dense layer=0 in your case, similar to Llama4.

File: src/maxtext/utils/maxtext_utils.py [hunk 2]
```diff
@@ -389,6 +389,77 @@ def get_dense_moe_layers(config):
...
+
+  # - Inter-chunk terms (Recurrent State D_k * D_v):
+  #   - attn_i (Q*K): 2 * B * S * H_v * C * D_k
+  #   - v_prime (K*S): 2 * B * S * H_v * D_k * D_v
+  #   - attn_inter (Q*S): 2 * B * S * H_v * D_k * D_v
+  #   - core_out (A*V): 2 * B * S * H_v * C * D_v
+  #   - update (K*V): 2 * B * S * H_v * D_k * D_v
+  flops_inter = (2 * B * S * H_v * C * (D_k + D_v)) + (6 * B * S * H_v * D_k * D_v)
+
+  flops_core = flops_intra + flops_inter
```
Comment (RissyRan) [inline src/maxtext/utils/maxtext_utils.py:557]: Do you think we should apply causal mask similar [here](https://github.com/AI-Hypercomputer/maxtext/blob/4bcee99601ae152ca8a9b642f1d597b17d769547/src/MaxText/maxtext_utils.py#L566-L570) or anywhere inside of `config.decoder_block == DecoderBlockType.QWEN3_NEXT`?

File: src/maxtext/utils/maxtext_utils.py [hunk 3]
```diff
@@ -389,6 +389,77 @@ def get_dense_moe_layers(config):
...
+  flops_out = 2 * B * S * V_dim * E
+
+  flops_projections = flops_qkvz + flops_ba + flops_out
+
+  # 2. Convolution (Learnable Weights)
+  # Depthwise conv on dim (2*K_dim + V_dim)
+  # 2 * B * S * Channels * Kernel
+  flops_conv = 2 * B * S * (2 * K_dim + V_dim) * K_conv
+
+  # 3. Core Gated Delta Net (Attention-like operations)
```
Comment (RissyRan) [inline src/maxtext/utils/maxtext_utils.py:531]: Thanks for the explanation! Could you help specify the shape of K, A, V, Q?

Comment (RissyRan) [reply src/maxtext/utils/maxtext_utils.py:531]: Also take a look at @parambole 's roof line analysis, it seems this part is dis-aligned?

https://screenshot.googleplex.com/BXZiCaUngeTe5rT

File: src/MaxText/maxtext_utils.py (hunk 1 above)
Comment (Rohan-Bierneni) [reply src/MaxText/maxtext_utils.py]: Sure, I changed the code to use calculate_routed_and_shared_ffn_tflops_per_device and removed calculate_qwen3_next_ffn_tflops_per_device.

File: src/maxtext/utils/maxtext_utils.py (hunk 2 above)
Comment (Rohan-Bierneni) [reply src/maxtext/utils/maxtext_utils.py:557]: I don't think we should apply the causal mask factor here. Unlike standard attention where we conceptually calculate an 2-D matrix and mask half of it, here we are performing a fixed-size state update sequentially for every chunk. Since we execute this operation fully for every step to maintain the recurrent state, these FLOPs are 'dense' and shouldn't be halved. 

However, note that in calculate_tflops_training_per_device, we do apply the causal mask reduction to the Full Attention layers

**Comment by Rohan-Bierneni** [comment]
> Thanks for the update!
> 
> We have some unit [tests](tests/flop_calculation_test.py) for those tflops, if you could come up with a test, that would be great, but not mandatory for this PR. Also, could you also have some sanity check with other MoE model (no change impact)? Similar to #2979

I have added a unit test in flop_calculation_test.py. Also will run the sanity checks and update the pr description with results.

File: src/maxtext/utils/maxtext_utils.py (hunk 3 above)
Comment (Rohan-Bierneni) [reply src/maxtext/utils/maxtext_utils.py:531]: Sure, the shapes are as follow:

Query (Q): [Batch, Seq_Len, H_v, D_k]
Keys (K): [Batch, Seq_Len, H_v, D_k]
Values (V): [Batch, Seq_Len, H_v, D_v]
Intra-Chunk Attention (A): [Batch, Num_Chunks, H_v, Chunk_Size, Chunk_Size]
Recurrent State (S): [Batch, Num_Chunks, H_v, D_k, D_v]

Comment (Rohan-Bierneni) [reply src/maxtext/utils/maxtext_utils.py:531]: > Also take a look at @parambole 's roof line analysis, it seems this part is dis-aligned?
> 
> https://screenshot.googleplex.com/BXZiCaUngeTe5rT

The roofline analysis accounts for the State Expansion & State Retrieval but not for the State Reconstruction step as part of the delta rule: https://github.com/AI-Hypercomputer/maxtext/blob/c63692493712e20c7b1a4b0c945115e1737177c6/src/MaxText/layers/qwen3.py#L234

This would only increase the tflops in the roofline analysis for core linear attention compute from 0.004 to 0.006 which is a pretty negligible amount when compared to tflops for the projections at 0.206 tflops.

File: src/maxtext/utils/maxtext_utils.py [hunk 4]
```diff
@@ -380,15 +380,74 @@ def get_dense_moe_layers(config):
...
+    num_moe_layers = config.num_decoder_layers // interleave_step
     num_dense_layers = config.num_decoder_layers - num_moe_layers
   else:
-    raise ValueError("Currently we only support DeepSeek and Llama4 calculation.")
+    raise ValueError("Currently we only support DeepSeek, Llama4, ad Qwen3-Next calculation.")
 
   return num_dense_layers, num_moe_layers
 
 
+def calculate_gated_delta_net_flops_per_device(config):
```
Comment (parambole) [inline src/maxtext/utils/maxtext_utils.py:500]: The current `flops_intra` calculation:

`flops_intra = 2 * B * S * H_v * C * (2 * D_k + D_v)`

This correctly captures the FLOPs from the three main `jnp.matmul` operations using the *final* `attn` matrix:
1.  Initial `attn` matrix: `jnp.matmul(k_beta_c, key_c^T)`
2.  `value_intra = jnp.matmul(attn, v_beta_c)`
3.  `k_cumdecay = jnp.matmul(attn, k_beta_c * ...)`

However, this calculation is **missing the significant computational cost of the `inner_attn_body` `fori_loop`** within the `jax_chunk_gated_delta_rule` function. This loop iteratively refines the `attn` matrix *before* the `value_intra` and `k_cumdecay` matrix multiplications.

The dominant operation within each of the `C-1` iterations of the loop is `jnp.sum(term, axis=-2)`, where `term` is `(B, H, N, C, C)`. The FLOPs for this loop are approximately:

`F_inner_loop ≈ (C - 1) * B * H * N * C^2 ≈ B * H * S * C^2`

This cost should be added to `flops_intra`. The more accurate `flops_intra` would be:

`flops_intra = 2 * B * S * H_v * C * (2 * D_k + D_v) + B * H_v * S * C^2`

Comment (Rohan-Bierneni) [reply src/maxtext/utils/maxtext_utils.py:500]: Thank you for catching this! I have updated the flops calculation to the right flops_intra formula

File: src/maxtext/utils/maxtext_utils.py (hunk 3 above)
Comment (Rohan-Bierneni) [reply src/maxtext/utils/maxtext_utils.py:531]: The core attention flops should be correct now, and I will update the roofline analysis as well to align on both tflops calculations.

File: src/MaxText/maxtext_utils.py [hunk 5]
```diff
@@ -380,15 +380,75 @@ def get_dense_moe_layers(config):
     num_dense_layers = config.first_num_dense_layers
     num_moe_layers = config.num_decoder_layers - config.first_num_dense_layers
     return num_dense_layers, num_moe_layers
-  elif config.decoder_block == DecoderBlockType.LLAMA4:
-    num_moe_layers = config.num_decoder_layers // config.interleave_moe_layer_step
+  elif config.decoder_block in (DecoderBlockType.LLAMA4, DecoderBlockType.QWEN3_NEXT):
```
Comment (RissyRan) [inline src/MaxText/maxtext_utils.py]: Could you help split those 2 blocks, add something like 

```
elif config.decoder_block == DecoderBlockType.QWEN3_NEXT:
  num_moe_layers = config.num_decoder_layers # all moe layers
  num_dense_layers = 0
```

File: src/MaxText/maxtext_utils.py [hunk 6]
```diff
@@ -380,15 +380,75 @@ def get_dense_moe_layers(config):
...
+  flops_projections = flops_qkvz + flops_ba + flops_out
+
+  # 2. Convolution (Learnable Weights)
+  # Depthwise conv on dim (2*K_dim + V_dim)
+  # 2 * B * S * Channels * Kernel
+  flops_conv = 2 * B * S * (2 * K_dim + V_dim) * K_conv
+
+  # 3. Core Gated Delta Net (Attention-like operations)
+  # Assumptions:
+  # - H = H_v (broadcasting K to V heads if H_v > H_k)
```
Comment (RissyRan) [inline src/MaxText/maxtext_utils.py]: This is really helpful comment for readers, could you help add here?

```
Query (Q): [Batch, Seq_Len, H_v, D_k]
Keys (K): [Batch, Seq_Len, H_v, D_k]
Values (V): [Batch, Seq_Len, H_v, D_v]
Intra-Chunk Attention (A): [Batch, Num_Chunks, H_v, Chunk_Size, Chunk_Size]
Recurrent State (S): [Batch, Num_Chunks, H_v, D_k, D_v]
```

File: src/maxtext/utils/maxtext_utils.py [hunk 7]
```diff
@@ -380,15 +380,75 @@ def get_dense_moe_layers(config):
...
+
+  # 3. Core Gated Delta Net (Attention-like operations)
+  # Assumptions:
+  # - H = H_v (broadcasting K to V heads if H_v > H_k)
+  # - Intra-chunk terms (per chunk C):
+  #   - attn (K*K): 2 * B * S * H_v * C * D_k
+  #   - val_intra (A*V): 2 * B * S * H_v * C * D_v
+  #   - k_cum (A*K): 2 * B * S * H_v * C * D_k
+  #   - inner_attn_body loop (iterative refinement): ≈ (C - 1) * B * H * N * C^2 ≈ B * H * S * C^2
+  flops_intra = 2 * B * S * H_v * C * (2 * D_k + D_v) + (B * H_v * S * C**2)
```
Comment (RissyRan) [inline src/maxtext/utils/maxtext_utils.py:547]: Sorry I didn't follow here when comparing with Param's doc. I understand the values are similar, but could we align the formula as well? I sent a quick sync-up to move it faster. Thank you!

```
     State Expansion (K^T x V): For every token, we compute the outer product.
       Math: [16, 4096, 128, 1] @ [16, 4096, 1, 128]
       FLOPs: 2 x 16 x 4096 x 128 x 128 / 10^12 = 0.002 TFLOPS
     State Retrieval (Q x S): For every token, we query the state.
       Math: [16, 4096, 1, 128] @ [16, 4096, 128, 128]
       FLOPs: 2 x 16 x 4096 x 128 x 128 / 10^12 = 0.002 TFLOPS
     Total: 0.004 TFLOPS

```

**Review by RissyRan — COMMENTED** [review]
Thanks for the change! I sent a quick sync-up, hopefully this could iterate faster.

File: src/MaxText/maxtext_utils.py (hunk 5 above)
Comment (Rohan-Bierneni) [reply src/MaxText/maxtext_utils.py]: Resolved.

File: src/MaxText/maxtext_utils.py (hunk 6 above)
Comment (Rohan-Bierneni) [reply src/MaxText/maxtext_utils.py]: I have added these to the comments for easier reading.

File: src/maxtext/utils/maxtext_utils.py (hunk 7 above)
Comment (Rohan-Bierneni) [reply src/maxtext/utils/maxtext_utils.py:547]: Based on conversation from internal sync, we will continue with the formulas in this PR, as @parambole's doc highlights the main compute being done, while the formulas in this PR is more specific/accurate to the GatedDeltaNet implementation. However, note that the delta between the 2 Tflop calculations differ only by 0.1% since input projections dominate the overall Tflops.

File: src/MaxText/maxtext_utils.py [hunk 8]
```diff
@@ -383,12 +383,82 @@ def get_dense_moe_layers(config):
   elif config.decoder_block == DecoderBlockType.LLAMA4:
     num_moe_layers = config.num_decoder_layers // config.interleave_moe_layer_step
     num_dense_layers = config.num_decoder_layers - num_moe_layers
+  elif config.decoder_block == DecoderBlockType.QWEN3_NEXT:
+    num_moe_layers = config.num_decoder_layers
+    num_dense_layers = 0
   else:
-    raise ValueError("Currently we only support DeepSeek and Llama4 calculation.")
+    raise ValueError("Currently we only support DeepSeek, Llama4, ad Qwen3-Next calculation.")
```
[the remark this answers is not in the document]
Comment (Rohan-Bierneni) [reply src/MaxText/maxtext_utils.py]: Resolved.

**Comment by Rohan-Bierneni** [comment]
Manually adding pull_ready tag since all tests passed and 2 approvals recieved
